File: dashboard/src/lib/roles.ts
export const ROLE_HIERARCHY = ['guest', 'viewer', 'member', 'admin', 'owner'] as const
export type Role = typeof ROLE_HIERARCHY[number]

export function hasMinRole(userRole: Role, minRole: Role): boolean {
  return ROLE_HIERARCHY.indexOf(userRole) >= ROLE_HIERARCHY.indexOf(minRole)
}

export function canAssignRole(assignerRole: Role, targetRole: Role): boolean {
  return ROLE_HIERARCHY.indexOf(assignerRole) >= ROLE_HIERARCHY.indexOf(targetRole)
}

export const ROLE_LABELS: Record<Role, string> = {
  owner: 'Owner',
  admin: 'Admin',
  member: 'Member',
  viewer: 'Viewer',
  guest: 'Guest',
}

export const ROLE_COLORS: Record<Role, string> = {
  owner: 'text-amber-400 bg-amber-400/10',
  admin: 'text-teal-400 bg-teal-400/10',
  member: 'text-stone-300 bg-stone-300/10',
  viewer: 'text-stone-500 bg-stone-500/10',
  guest: 'text-stone-600 bg-stone-600/10',
}

File: orchestrator/app/roles.py
"""Fixed role definitions and hierarchy for RBAC."""
from enum import IntEnum


class Role(IntEnum):
    GUEST = 0
    VIEWER = 1
    MEMBER = 2
    ADMIN = 3
    OWNER = 4


ROLE_NAMES = {r.name.lower(): r for r in Role}
VALID_ROLES = set(ROLE_NAMES.keys())


def parse_role(role_str: str) -> Role:
    return ROLE_NAMES.get(role_str.lower(), Role.MEMBER)


def can_assign_role(assigner_role: str, target_role: str) -> bool:
    return parse_role(assigner_role) >= parse_role(target_role)


def has_min_role(user_role: str, min_role: str) -> bool:
    return parse_role(user_role) >= parse_role(min_role)
